fixing pandas.DataFrame.plot(): labels do not appear in legend and label kwd

schmohlio opened the pull request:
Closes https://github.com/pydata/pandas/issues/9542, #8905

The following behavior has been tested: 
- `df.plot(y='sin(x)')` -> gives a legend with label 'None' -> this should give no legend instead (as it plots one series, and then we don't automatically add a legend, see behaviour of df['sin(x)'].plot())
- `df.plot(y='sin(x)', legend=True)` -> gives a legend with label 'None' -> this should of course give a legend with label 'sin(x)' (behaviour as df['sin(x)'].plot(legend=True))
- `df.plot(y='sin(x)', label='something else', legend=True)` -> gives a legend with label 'None' -> should be a legend with label 'something else', as we want that the label kwarg overwrites the column name.

based on following data:
`x=np.linspace(-10,10,201)`
`y,z=np.sin(x),np.cos(x)`
`x,y,z=pd.Series(x),pd.Series(y),pd.Series(z)`
`df=pd.concat([x,y,z],axis=1)`
`df.columns=['x','sin(x)','cos(x)']`
`df=df.set_index('x')`

Commit:
fixing pandas.DataFrame.plot(): Labels do not appear in Legend

schmohlio commented:
not sure why there would be a test_url failure in the build...

jorisvandenbossche commented:
@schmohlio I think you can ignore that test_url error, it does not seem to be related to this PR.

You can indeed put the tests in `test_graphics`. 
I did not yet look in detail, but a quick feedback: I don't think it should be necessary to change the default value for `legend`. Further, is it needed to change the existing tests?

cc @TomAugspurger @sinhrks

schmohlio commented:
Thanks for looking @jorisvandenbossche 

Is there an elegant way to check for user supplied arg (legend=) when default is not None?

I didn't spend as much time on tests as I've like to yet, but I made changes to align plot_frame with plot_series. Previously, label wasn't getting passed through at all and was overwriting index name within plot_frame. I could potentially create logic for label that changes the axis only if the number of series is 1 to align with old test, but this wouldn't match plot_series per our tests above. Will await more feedback; thanks again.

schmohlio commented:
also, the tests changed per the original issue; it probably doesn't make sense to mutate the index name with `label=`

TomAugspurger commented on pandas/tools/plotting.py:
```diff
@@ -886,10 +891,11 @@ def _iter_data(self, data=None, keep_index=False, fillna=None):
 
         from pandas.core.frame import DataFrame
         if isinstance(data, (Series, np.ndarray, Index)):
+            label = self.label if self.label else data.name
```
Need to be careful here. We need to treat `False` differently than `None`, in case someone wants the actual label in the legend to be the text "False". do `label = self.label if self.label is not None else data.name`.

TomAugspurger commented:
I'll try to look more later.

schmohlio commented:
Good catch.

TomAugspurger commented on pandas/tools/plotting.py:
```diff
@@ -816,7 +816,12 @@ def __init__(self, data, kind=None, by=None, subplots=False, sharex=True,
 
         self.grid = grid
-        self.legend = legend
+        if legend is not None:
```
Could you explain this if block? I'm not sure I see the point of it.

schmohlio replied:
The original point of this was decide if a label should be rendered based on whether there was a valid label provided or if there was more than 1 series. I can revert back the defaults, though, and make legend show default series names of no legend is provided.

TomAugspurger commented:
I agree with Joris that the default for legend doesn't need to change.

TomAugspurger commented on pandas/tests/test_graphics.py:
```diff
@@ -2518,12 +2512,12 @@ def test_df_legend_labels(self):
 
         # scatter
-        ax = df.plot(kind='scatter', x='a', y='b', label='data1')
+        ax = df.plot(kind='scatter', x='a', y='b', label='data1', legend=True)
```
When you change the default for legend back to True you can remove the `legend=True`s that you added.

schmohlio replied:
Sounds good.

TomAugspurger commented on pandas/tests/test_graphics.py:
```diff
@@ -1065,12 +1065,6 @@ def test_implicit_label(self):
         self._check_text_labels(ax.xaxis.get_label(), 'a')
 
     @slow
-    def test_explicit_label(self):
```
Can you add this test back in, or say why you removed it? I'd guess people would rely on specifying label here implies a legend (even though it doesn't with Series.plot :frown: )

schmohlio replied:
Hey @TomAugspurger - this test checks to see if th axis is relabeled and not if there is a legend title label. Per the issue discussion we thought the axis shouldn't be changed, but just the series name within the legend. There is another test that checks legend that I edited but can re-edit

jorisvandenbossche replied:
It is indeed possible people rely on this, but it *is* very inconsistent, so strange it is put into a test. `label` clearly should not alter the 'xlabel', as it is used in matplotlib also for the legends

schmohlio replied:
yes and this test doesn't specify a legend, rather the index name

TomAugspurger replied:
Sorry! I didn't realize the test was for the *axis* label. Good call on removing it.

schmohlio commented:
@jorisvandenbossche @TomAugspurger - here is a summary. I'm going to do the following:

1. Revert to default `legend=True`. 
  * this means that even if no `label=` arg is provided by the user and only 1 series is plotted, the legend will still show something based on #2. This will just be a difference for Series and DataFrames. 
 * Previously I had introduced logic based on the number of series.
 * will revert back `legend=True` on test edits
2. Ensure the following logic for coalescing the legend labels:
  * check `label=` arg (and make sure it is passed through)--> check series name property (y label) --> check for index name 
3. Keep index name test removed, and keep other tests that check legend. Maybe expand upon these.

TomAugspurger commented:
Thanks for the summary. That all looks perfect.

schmohlio commented:
I added tests to the proper check legend labels method. Altogether, this test method validates all the changes we discussed. Now the `label=` arg functions intuitively. I also added a check to make sure there is no column mutation. Thanks for patience.

Pushing now @TomAugspurger @jorisvandenbossche

TomAugspurger commented:
Great thanks. If you could just rebase + squash and add a release not I'll get it merged.

schmohlio commented:
Hey @TomAugspurger Tom - just squashed my commits, rebased, pulled, and pushed. Hope that's okay.

jorisvandenbossche commented:
Is it possible to keep the behaviour as before? So that when plotting one series by specifying the `y` kwarg no legend is shown? (although I don't know if this is the best behaviour, but when plotting one series, it looks to me as the name of the series should be the `ylabel`)

schmohlio commented:
@jorisvandenbossche - right now a legend will appear with the ylabel as the label, unless the user explicitly supplies legend=False. I agreed before but seeing it in action now it almost seems intuitive to have the legend default to true. If we are going to keep the default to True it seems a little much to override it somehow in case the user wants legend.

schmohlio commented:
The tests also line up. What if we added another ticket for adding a legend to series plots?

schmohlio commented:
@TomAugspurger I added release notes and squashed my commits properly. Figured we could go from there in a cleaner manner.

TomAugspurger commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -546,6 +546,37 @@ Bug Fixes
 - Bug in ``rank`` where comparing floats with tolerance will cause inconsistent behaviour (:issue:`8365`).
 - Fixed character encoding bug in ``read_stata`` and ``StataReader`` when loading data from a URL (:issue:`9231`).
 - Bug in adding ``offsets.Nano`` to other offets raises ``TypeError`` (:issue:`9284`)
+<<<<<<< HEAD
```
Thanks for rebasing. You can remove this line, the one below it, and the one on 579. git puts these in when the same line has been edited on your branch and what you're rebasing onto.

TomAugspurger commented on doc/source/whatsnew/v0.16.0.txt:
```diff
@@ -546,6 +546,37 @@ Bug Fixes
+
+- Fixed bug (:issue:`9542`) where labels did not appear in legend (instead, ``None`` types were displayed for each series) when passing ``label=`` args to ``dataframe.plot()``. Also, series indices are no longer mutated and labels are default to column names if ``label=`` args are not provided. Use ``legend=False`` to avoid printing a legend.
+
+  Fixed behavior:
```
This could be useful in the first post of the linked issue. Usually we keep the release notes brief so people can quickly scan through them. If they see something interesting they can click-through to the issue or pull request. Great detail though!

schmohlio replied:
cool i added this to the issue (not sure how to bump it to the top). Thanks @TomAugspurger

TomAugspurger commented:
I think we're really close. Code-wise everything looks good. Just those couple things in the release notes. If you want to just rebase *one* more time :)

I'll be in our chat room at https://gitter.im/pydata/pandas Give me a shout if you have any trouble.

TomAugspurger commented:
And @schmohlio looks like you'll be one of the first ones in `0.16.1`. Could you make a new file at `doc/source/whatsnew/v0.16.1.txt` And paste this in:

```
.. _whatsnew_0161:

v0.16.1 (2015)
-----------------------

This is a minor release from 0.16.0 and includes a small number of API changes, several new features,
enhancements, and performance improvements along with a large number of bug fixes. We recommend that all
users upgrade to this version.


- :ref:`Enhancements <whatsnew_0161.enhancements>`
- :ref:`API Changes <whatsnew_0161.api>`
- :ref:`Performance Improvements <whatsnew_0161.performance>`
- :ref:`Experimental Changes <whatsnew_0161.experimental>`
- :ref:`Bug Fixes <whatsnew_0161.bug_fixes>`

.. _whatsnew_0161.api:

API changes
~~~~~~~~~~~

.. _whatsnew_0161.enhancements:

Enhancements
~~~~~~~~~~~~




.. _whatsnew_0161.performance:

Performance
~~~~~~~~~~~





.. _whatsnew_0161.experimental:

Experimental
~~~~~~~~~~~~





.. _whatsnew_0161.bug_fixes:

Bug Fixes
~~~~~~~~~
```

Then you can move your section in `doc/source/whatsnew/v0.16.0` to that new file. Thanks.

schmohlio commented:
cool this looks good. will shorten and rebase with new template.

jreback commented:
whats new is now available for 0.16.1 https://github.com/pydata/pandas/commit/5ebf521eac528f5a4b772e4677f70b689b8b1826

schmohlio commented:
@jreback should i remove the whatsnew file I had added?

jreback commented:
yes, when you pull and rebase you will get the new one

schmohlio commented:
@jreback merged in whats new

TomAugspurger commented:
Thanks @schmohlio!

Looks like this was your first PR for pandas. Stick around if you find some more bugs :)

schmohlio commented:
@TomAugspurger thanks to you! will look out for sure.
